README.rst
swh-icinga-plugins
==================

Scripts for end-to-end monitoring of the SWH infrastructure


debian/changelog
swh.icingaplugins (0.6.0-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.6.0     - (tagged by David Douard
    <david.douard@sdfa3.org> on 2023-12-05 11:52:22 +0100)
  * Upstream changes:     - v0.6.0     - Migrate to copier, PEP420 and
    pyproject.toml based packaging.

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Tue, 05 Dec 2023 10:55:43 +0000

swh.icingaplugins (0.5.6-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.5.6     - (tagged by Valentin Lorentz
    <vlorentz@softwareheritage.org> on 2023-11-28 11:54:12 +0100)
  * Upstream changes:     - v0.5.6     - * jesuisgpl.tgz.xml: Remove
    elements with invalid tags

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Tue, 28 Nov 2023 10:57:55 +0000

swh.icingaplugins (0.5.5-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.5.5     - (tagged by Antoine R. Dumont
    (@ardumont) <ardumont@softwareheritage.org> on 2023-11-27 14:56:41
    +0100)
  * Upstream changes:     - v0.5.5     - deposit: Deal with rate limit
    during get queries     - docs: include the README file in the main
    index page

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Mon, 27 Nov 2023 14:00:16 +0000

swh.icingaplugins (0.5.4-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.5.4     - (tagged by Nicolas Dandrimont
    <nicolas@dandrimont.eu> on 2023-10-30 13:19:15 +0100)
  * Upstream changes:     - Release swh.icinga.plugins v0.5.4     -
    Support modern vault deployment

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Mon, 30 Oct 2023 12:22:26 +0000

swh.icingaplugins (0.5.3-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.5.3     - (tagged by Nicolas Dandrimont
    <nicolas@dandrimont.eu> on 2023-01-18 15:24:35 +0100)
  * Upstream changes:     - Release swh.icinga_plugins 0.5.3     - Hook
    up multi-origin save code now check to the CLI     - Ensure vault
    tests work on Python3.11     - spruce up formatting

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Wed, 18 Jan 2023 14:28:09 +0000

swh.icingaplugins (0.5.2-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.5.2     - (tagged by Valentin Lorentz
    <vlorentz@softwareheritage.org> on 2022-09-27 14:15:05 +0200)
  * Upstream changes:     - v0.5.2     - * SaveCodeNowCheck: Add support
    for list of origins

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Tue, 27 Sep 2022 12:19:10 +0000

swh.icingaplugins (0.5.1-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.5.1     - (tagged by Vincent SELLIER
    <vincent.sellier@softwareheritage.org> on 2022-05-10 18:11:41 +0200)
  * Upstream changes:     - v0.5.1

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Tue, 10 May 2022 16:15:00 +0000

swh.icingaplugins (0.5.0-1~swh2) unstable-swh; urgency=medium

  * add a missing dependency 

 -- Vincent Sellier <vincent.sellier@softwareheritage.org>  Tue, 10 May 2022 14:14:10 +0200

swh.icingaplugins (0.5.0-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.5.0     - (tagged by Vincent SELLIER
    <vincent.sellier@softwareheritage.org> on 2022-05-10 09:16:08 +0200)
  * Upstream changes:     - v0.5.0     - some changes in the tests     -
    add prometheus metrics

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Tue, 10 May 2022 07:20:03 +0000

swh.icingaplugins (0.4.2-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.4.2     - (tagged by Valentin Lorentz
    <vlorentz@softwareheritage.org> on 2022-03-04 19:00:58 +0100)
  * Upstream changes:     - v0.4.2     - * deposit: Properly escape URLs
    - * deposit: Check the status code of the REMD API.

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Fri, 04 Mar 2022 18:03:24 +0000

swh.icingaplugins (0.4.1-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.4.1     - (tagged by Valentin Lorentz
    <vlorentz@softwareheritage.org> on 2022-03-04 18:18:17 +0100)
  * Upstream changes:     - v0.4.1     - * vault: Open tarballs in
    stream mode + add test for empty tar files     - * deposit: Fetch
    only metadata created after the check started

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Fri, 04 Mar 2022 17:21:14 +0000

swh.icingaplugins (0.4.0-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.4.0     - (tagged by Nicolas Dandrimont
    <nicolas@dandrimont.eu> on 2022-03-04 17:26:11 +0100)
  * Upstream changes:     - Release swh-icinga-plugins 0.4.0     - more
    vault checks     - cleaner deposit checks

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Fri, 04 Mar 2022 16:28:49 +0000

swh.icingaplugins (0.3.2-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.3.2     - (tagged by Nicolas Dandrimont
    <nicolas@dandrimont.eu> on 2021-06-18 15:29:50 +0200)
  * Upstream changes:     - Release swh.icinga_plugins 0.3.2     - Don't
    require type stub modules at runtime

 -- Software Heritage autobuilder (on jenkins-debian1) <jenkins@jenkins-debian1.internal.softwareheritage.org>  Fri, 18 Jun 2021 13:34:21 +0000

swh.icingaplugins (0.3.1-1~swh1) unstable-swh; urgency=medium

  * New upstream release 0.3.1     - (tagged by Nicolas Dandrimont
    <nicolas@dandrimont.eu> on 2021-06-18 14:39:11 +0200)
  * Upstream changes:     - Release swh.icinga.plugins v0.3.1     - Add

swh/icinga_plugins/tests/utils.py
# Copyright (C) 2019-2021  The Software Heritage developers
# See the AUTHORS file at the top-level directory of this distribution
# License: GNU General Public License version 3, or any later version
# See top-level LICENSE file for more information

from typing import List

from click.testing import CliRunner, Result

from swh.icinga_plugins.cli import icinga_cli_group


def invoke(args: List[str], catch_exceptions: bool = False) -> Result:
    """Invoke icinga plugin main cli command with args"""
    runner = CliRunner()
    result = runner.invoke(icinga_cli_group, args)
    if not catch_exceptions and result.exception:
        print(result.output)
        raise result.exception
    return result


swh/icinga_plugins/tests/conftest.py
# Copyright (C) 2019-2022  The Software Heritage developers
# See the AUTHORS file at the top-level directory of this distribution
# License: GNU General Public License version 3, or any later version
# See top-level LICENSE file for more information

import pytest


@pytest.fixture
def mocked_time(mocker):
    start_time = 1646413359.0  # 2022-03-04-17:02:39Z

    time_offset = 0

    def fake_sleep(seconds):
        nonlocal time_offset
        time_offset += seconds

    def fake_time():
        return start_time + time_offset

    mocker.patch("time.sleep", side_effect=fake_sleep)
    mocker.patch("time.time", side_effect=fake_time)


swh/icinga_plugins/tests/test_vault.py
# Copyright (C) 2019-2021  The Software Heritage developers
# See the AUTHORS file at the top-level directory of this distribution
# License: GNU General Public License version 3, or any later version
# See top-level LICENSE file for more information

import io
import sys
import tarfile
import time

from swh.icinga_plugins.tests.utils import invoke

from .web_scenario import WebScenario

DIR_ID = "ab" * 20

url_api = f"mock://swh-web.example.org/api/1/vault/directory/{DIR_ID}/"
url_fetch = f"mock://swh-web.example.org/api/1/vault/directory/{DIR_ID}/raw/"
url_fetch_redirected = (
    f"mock://swh-web-redirected.example.org/api/1/vault/directory/{DIR_ID}/raw/"
)


def _make_tarfile():
    fd = io.BytesIO()
    with tarfile.open(fileobj=fd, mode="w:gz") as tf:
        tf.addfile(tarfile.TarInfo(f"swh:1:dir:{DIR_ID}/README"), b"this is a readme\n")

        tarinfo = tarfile.TarInfo(f"swh:1:dir:{DIR_ID}")
        tarinfo.type = tarfile.DIRTYPE
        tf.addfile(tarinfo)
    return fd.getvalue()


TARBALL = _make_tarfile()

response_pending = {
    "obj_id": DIR_ID,
    "obj_type": "directory",
    "progress_message": "foo",
    "status": "pending",
}

response_done = {
    "fetch_url": url_fetch,
    "id": 9,
    "obj_id": DIR_ID,
    "obj_type": "directory",
    "status": "done",
}

response_done_no_fetch = {
    "id": 9,
    "obj_id": DIR_ID,
    "obj_type": "directory",
    "status": "done",
}

response_failed = {
    "obj_id": DIR_ID,
    "obj_type": "directory",
    "progress_message": "foobar",
    "status": "failed",
}

response_unknown_status = {
    "obj_id": DIR_ID,
    "obj_type": "directory",
    "progress_message": "what",
    "status": "boo",
}


class FakeStorage:
    def __init__(self, foo, **kwargs):
        pass

    def directory_get_random(self):
        return bytes.fromhex(DIR_ID)


def test_vault_immediate_redirect(requests_mock, mocker, mocked_time):
    scenario = WebScenario()

    scenario.add_step("get", url_api, {}, status_code=404)
    scenario.add_step("post", url_api, response_pending)
    scenario.add_step("get", url_api, response_done)
    scenario.add_step(
        "get",
        url_fetch,
        {},
        status_code=302,
        headers={"Location": url_fetch_redirected},
    )
    scenario.add_step(
        "get",
        url_fetch_redirected,
        TARBALL,
        headers={"Content-Type": "application/octet-stream"},
    )

    scenario.install_mock(requests_mock)

    get_storage_mock = mocker.patch("swh.icinga_plugins.vault.get_storage")
    get_storage_mock.side_effect = FakeStorage

    result = invoke(
        [
            "check-vault",
            "--swh-web-url",
            "mock://swh-web.example.org",
            "--swh-storage-url",
            "foo://example.org",
            "directory",
        ]
    )

    assert result.output == (
        f"VAULT OK - cooking directory {DIR_ID} took "
        f"10.00s and succeeded.\n"
        f"| 'total_time' = 10.00s\n"
    )
    assert result.exit_code == 0, result.output


def test_vault_immediate_success(requests_mock, mocker, mocked_time):
    scenario = WebScenario()

    scenario.add_step("get", url_api, {}, status_code=404)
    scenario.add_step("post", url_api, response_pending)
    scenario.add_step("get", url_api, response_done)
    scenario.add_step(
        "get", url_fetch, TARBALL, headers={"Content-Type": "application/gzip"}
    )

    scenario.install_mock(requests_mock)

    get_storage_mock = mocker.patch("swh.icinga_plugins.vault.get_storage")
    get_storage_mock.side_effect = FakeStorage

    result = invoke(
        [
            "check-vault",
            "--swh-web-url",
            "mock://swh-web.example.org",
            "--swh-storage-url",
            "foo://example.org",
            "directory",
        ]
    )

    assert result.output == (
        f"VAULT OK - cooking directory {DIR_ID} took "
        f"10.00s and succeeded.\n"
        f"| 'total_time' = 10.00s\n"
    )
    assert result.exit_code == 0, result.output


def test_vault_delayed_success(requests_mock, mocker, mocked_time):
    scenario = WebScenario()

    scenario.add_step("get", url_api, {}, status_code=404)
    scenario.add_step("post", url_api, response_pending)
    scenario.add_step("get", url_api, response_pending)
    scenario.add_step("get", url_api, response_done)
    scenario.add_step(
        "get", url_fetch, TARBALL, headers={"Content-Type": "application/gzip"}
    )

    scenario.install_mock(requests_mock)

    get_storage_mock = mocker.patch("swh.icinga_plugins.vault.get_storage")
    get_storage_mock.side_effect = FakeStorage

    result = invoke(
        [
            "check-vault",
            "--swh-web-url",
            "mock://swh-web.example.org",
            "--swh-storage-url",
            "foo://example.org",
            "directory",
        ]
    )

    assert result.output == (
        f"VAULT OK - cooking directory {DIR_ID} took "
        f"20.00s and succeeded.\n"
        f"| 'total_time' = 20.00s\n"
    )
    assert result.exit_code == 0, result.output


def test_vault_failure(requests_mock, mocker, mocked_time):
    scenario = WebScenario()

    scenario.add_step("get", url_api, {}, status_code=404)
    scenario.add_step("post", url_api, response_pending)
    scenario.add_step("get", url_api, response_failed)

    scenario.install_mock(requests_mock)

    get_storage_mock = mocker.patch("swh.icinga_plugins.vault.get_storage")
    get_storage_mock.side_effect = FakeStorage

    result = invoke(
        [
            "--prometheus-exporter",
            "--prometheus-exporter-directory",
            "/tmp",
            "check-vault",
            "--swh-web-url",
            "mock://swh-web.example.org",
            "--swh-storage-url",
            "foo://example.org",
            "directory",
        ],
        catch_exceptions=True,
    )

    assert result.output == (
        f"VAULT CRITICAL - cooking directory {DIR_ID} took "
        f"10.00s and failed with: foobar\n"
        f"| 'total_time' = 10.00s\n"
    )
    assert result.exit_code == 2, result.output


def test_vault_unknown_status(requests_mock, mocker, mocked_time):
    scenario = WebScenario()

    scenario.add_step("get", url_api, {}, status_code=404)
    scenario.add_step("post", url_api, response_pending)
    scenario.add_step("get", url_api, response_unknown_status)

    scenario.install_mock(requests_mock)

    get_storage_mock = mocker.patch("swh.icinga_plugins.vault.get_storage")
    get_storage_mock.side_effect = FakeStorage

    result = invoke(
        [
            "--prometheus-exporter",
            "--prometheus-exporter-directory",
            "/tmp",
            "check-vault",
            "--swh-web-url",
            "mock://swh-web.example.org",
            "--swh-storage-url",
            "foo://example.org",
            "directory",
        ],
        catch_exceptions=True,
    )

    assert result.output == (
        f"VAULT CRITICAL - cooking directory {DIR_ID} took "
        f"10.00s and resulted in unknown status: boo\n"
        f"| 'total_time' = 10.00s\n"
    )
    assert result.exit_code == 2, result.output


def test_vault_timeout(requests_mock, mocker, mocked_time):
    scenario = WebScenario()

    scenario.add_step("get", url_api, {}, status_code=404)
    scenario.add_step("post", url_api, response_pending)
    scenario.add_step("get", url_api, response_pending)
    scenario.add_step(
        "get", url_api, response_pending, callback=lambda: time.sleep(4000)
    )

    scenario.install_mock(requests_mock)

    get_storage_mock = mocker.patch("swh.icinga_plugins.vault.get_storage")
    get_storage_mock.side_effect = FakeStorage

    result = invoke(
        [
            "--prometheus-exporter",
            "--prometheus-exporter-directory",
            "/tmp",
            "check-vault",
            "--swh-web-url",
            "mock://swh-web.example.org",
            "--swh-storage-url",
            "foo://example.org",
            "directory",
        ],
        catch_exceptions=True,
    )

    assert result.output == (
        f"VAULT CRITICAL - cooking directory {DIR_ID} took more than "
        f"4020.00s and has status: foo\n"
        f"| 'total_time' = 4020.00s\n"
    )
    assert result.exit_code == 2, result.output


def test_vault_cached_directory(requests_mock, mocker, mocked_time):
    """First serves a directory that's already in the cache, to
    test that vault_check requests another one."""
    scenario = WebScenario()

    scenario.add_step("get", url_api, {}, status_code=200)
    scenario.add_step("get", url_api, {}, status_code=404)
    scenario.add_step("post", url_api, response_pending)
    scenario.add_step("get", url_api, response_done)
    scenario.add_step(
        "get", url_fetch, TARBALL, headers={"Content-Type": "application/gzip"}
    )

    scenario.install_mock(requests_mock)

    get_storage_mock = mocker.patch("swh.icinga_plugins.vault.get_storage")
    get_storage_mock.side_effect = FakeStorage

    result = invoke(
        [
            "--prometheus-exporter",
            "--prometheus-exporter-directory",
            "/tmp",
            "check-vault",
            "--swh-web-url",
            "mock://swh-web.example.org",
            "--swh-storage-url",
            "foo://example.org",
            "directory",
        ]
    )

    assert result.output == (
        f"VAULT OK - cooking directory {DIR_ID} took "
        f"10.00s and succeeded.\n"
        f"| 'total_time' = 10.00s\n"
    )
    assert result.exit_code == 0, result.output


def test_vault_no_directory(requests_mock, mocker, mocked_time):
    """Tests with an empty storage"""
    scenario = WebScenario()
    scenario.install_mock(requests_mock)

    get_storage_mock = mocker.patch("swh.icinga_plugins.vault.get_storage")
    get_storage_mock.side_effect = FakeStorage
    mocker.patch(f"{__name__}.FakeStorage.directory_get_random", return_value=None)

    result = invoke(
        [
            "--prometheus-exporter",
            "--prometheus-exporter-directory",
            "/tmp",
            "check-vault",
            "--swh-web-url",
            "mock://swh-web.example.org",
            "--swh-storage-url",
            "foo://example.org",
            "directory",
        ],
        catch_exceptions=True,
    )

    assert result.output == ("VAULT CRITICAL - No directory exists in the archive.\n")
    assert result.exit_code == 2, result.output


def test_vault_fetch_failed(requests_mock, mocker, mocked_time):
    scenario = WebScenario()

    scenario.add_step("get", url_api, {}, status_code=404)
    scenario.add_step("post", url_api, response_pending)
    scenario.add_step("get", url_api, response_done)
    scenario.add_step(
        "get",
        url_fetch,
        "",
        status_code=500,
        headers={"Content-Type": "application/gzip"},
    )

    scenario.install_mock(requests_mock)

    get_storage_mock = mocker.patch("swh.icinga_plugins.vault.get_storage")
    get_storage_mock.side_effect = FakeStorage

    result = invoke(
        [
            "check-vault",
            "--swh-web-url",
            "mock://swh-web.example.org",
            "--swh-storage-url",
            "foo://example.org",
            "directory",
        ],
        catch_exceptions=True,
    )

    assert result.output == (
        f"VAULT CRITICAL - cooking directory {DIR_ID} took "
        f"10.00s and succeeded, but fetch failed with status code 500.\n"
        f"| 'total_time' = 10.00s\n"
    )
    assert result.exit_code == 2, result.output


def test_vault_fetch_missing_content_type(requests_mock, mocker, mocked_time):
    scenario = WebScenario()

    scenario.add_step("get", url_api, {}, status_code=404)
    scenario.add_step("post", url_api, response_pending)
    scenario.add_step("get", url_api, response_done)
    scenario.add_step("get", url_fetch, "")

    scenario.install_mock(requests_mock)

    get_storage_mock = mocker.patch("swh.icinga_plugins.vault.get_storage")
    get_storage_mock.side_effect = FakeStorage

    result = invoke(
        [
            "check-vault",
            "--swh-web-url",
            "mock://swh-web.example.org",
            "--swh-storage-url",
            "foo://example.org",
            "directory",
        ],
        catch_exceptions=True,
    )

    assert result.output == (
        "VAULT CRITICAL - Unexpected Content-Type when downloading bundle: None\n"
        "| 'total_time' = 10.00s\n"
    )
    assert result.exit_code == 2, result.output


def test_vault_corrupt_tarball_gzip(requests_mock, mocker, mocked_time):
    scenario = WebScenario()

    scenario.add_step("get", url_api, {}, status_code=404)
    scenario.add_step("post", url_api, response_pending)
    scenario.add_step("get", url_api, response_pending)
    scenario.add_step("get", url_api, response_done)
    scenario.add_step(
        "get",
        url_fetch,
        b"this-is-not-a-tarball",
        headers={"Content-Type": "application/gzip"},
    )

    scenario.install_mock(requests_mock)

    get_storage_mock = mocker.patch("swh.icinga_plugins.vault.get_storage")
    get_storage_mock.side_effect = FakeStorage

    result = invoke(
        [
            "check-vault",
            "--swh-web-url",
            "mock://swh-web.example.org",
            "--swh-storage-url",
            "foo://example.org",
            "directory",
        ],
        catch_exceptions=True,
    )

    assert result.output == (
        "VAULT CRITICAL - ReadError while reading tarball: not a gzip file\n"
        "| 'total_time' = 20.00s\n"
    )
    assert result.exit_code == 2, result.output


def test_vault_corrupt_tarball_member(requests_mock, mocker, mocked_time):
    fd = io.BytesIO()
    with tarfile.open(fileobj=fd, mode="w:gz") as tf:
        tf.addfile(tarfile.TarInfo("wrong_dir_name/README"), b"this is a readme\n")
    tarball = fd.getvalue()

    scenario = WebScenario()

    scenario.add_step("get", url_api, {}, status_code=404)
    scenario.add_step("post", url_api, response_pending)
    scenario.add_step("get", url_api, response_pending)
    scenario.add_step("get", url_api, response_done)
    scenario.add_step(
        "get",
        url_fetch,
        tarball,
        headers={"Content-Type": "application/gzip"},
    )

    scenario.install_mock(requests_mock)

    get_storage_mock = mocker.patch("swh.icinga_plugins.vault.get_storage")
    get_storage_mock.side_effect = FakeStorage

    result = invoke(
        [
            "check-vault",
            "--swh-web-url",
            "mock://swh-web.example.org",
            "--swh-storage-url",
            "foo://example.org",
            "directory",
        ],
        catch_exceptions=True,
    )

    assert result.output == (
        "VAULT CRITICAL - Unexpected member in tarball: wrong_dir_name/README\n"
        "| 'total_time' = 20.00s\n"
    )
    assert result.exit_code == 2, result.output


def test_vault_empty_tarball(requests_mock, mocker, mocked_time):
    fd = io.BytesIO()
    with tarfile.open(fileobj=fd, mode="w:gz"):
        pass
    tarball = fd.getvalue()
    print(tarball)

    scenario = WebScenario()

    scenario.add_step("get", url_api, {}, status_code=404)
    scenario.add_step("post", url_api, response_pending)
    scenario.add_step("get", url_api, response_pending)
    scenario.add_step("get", url_api, response_done)
    scenario.add_step(
        "get",
        url_fetch,
        tarball,
        headers={"Content-Type": "application/gzip"},
    )

    scenario.install_mock(requests_mock)

    get_storage_mock = mocker.patch("swh.icinga_plugins.vault.get_storage")
    get_storage_mock.side_effect = FakeStorage

    result = invoke(
        [
            "check-vault",
            "--swh-web-url",
            "mock://swh-web.example.org",
            "--swh-storage-url",
            "foo://example.org",
            "directory",
        ],
        catch_exceptions=True,
    )

    assert result.exit_code == 2, result.output

    if sys.version_info >= (3, 11):
        assert result.output == (
            "VAULT CRITICAL - Fetched tarball is empty\n| 'total_time' = 20.00s\n"
        )
    else:
        assert result.output == (
            "VAULT CRITICAL - StreamError while reading tarball (empty file?): "
            "seeking backwards is not allowed\n"
            "| 'total_time' = 20.00s\n"
        )


def test_vault_no_fetch_url(requests_mock, mocker, mocked_time):
    scenario = WebScenario()

    scenario.add_step("get", url_api, {}, status_code=404)
    scenario.add_step("post", url_api, response_pending)
    scenario.add_step("get", url_api, response_done_no_fetch)

    scenario.install_mock(requests_mock)

    get_storage_mock = mocker.patch("swh.icinga_plugins.vault.get_storage")
    get_storage_mock.side_effect = FakeStorage

    result = invoke(
        [
            "check-vault",
            "--swh-web-url",
            "mock://swh-web.example.org",
            "--swh-storage-url",
            "foo://example.org",
            "directory",
        ],
        catch_exceptions=True,
    )

    assert result.output == (
        f"VAULT CRITICAL - cooking directory {DIR_ID} took 10.00s and succeeded, "
        f"but API response did not contain a fetch_url.\n"
        f"| 'total_time' = 10.00s\n"
    )
    assert result.exit_code == 2, result.output


swh/icinga_plugins/vault.py
# Copyright (C) 2019-2022  The Software Heritage developers
# See the AUTHORS file at the top-level directory of this distribution
# License: GNU General Public License version 3, or any later version
# See top-level LICENSE file for more information

import sys
import tarfile
import time
from typing import List

import requests

from swh.storage import get_storage

from .base_check import BaseCheck


class NoDirectory(Exception):
    pass


class VaultCheck(BaseCheck):
    TYPE = "VAULT"
    DEFAULT_WARNING_THRESHOLD = 0
    DEFAULT_CRITICAL_THRESHOLD = 3600

    def __init__(self, obj):
        super().__init__(obj, application="vault")
        self._swh_storage = get_storage("remote", url=obj["swh_storage_url"])
        self._swh_web_url = obj["swh_web_url"]
        self._poll_interval = obj["poll_interval"]

        self.register_prometheus_gauge("status", "")
        self.register_prometheus_gauge("duration", "seconds", ["step", "status"])

    def _url_for_dir(self, dir_id):
        return self._swh_web_url + f"/api/1/vault/directory/{dir_id.hex()}/"

    def _pick_directory(self):
        dir_ = self._swh_storage.directory_get_random()
        if dir_ is None:
            raise NoDirectory()
        return dir_

    def _pick_uncached_directory(self):
        while True:
            dir_id = self._pick_directory()
            response = requests.get(self._url_for_dir(dir_id))
            if response.status_code == 404:
                return dir_id

    def _collect_prometheus_metrics(
        self, status: int, duration: float, labels: List[str]
    ) -> None:
        self.collect_prometheus_metric("status", status)
        self.collect_prometheus_metric(
            "duration",
            duration,
            labels,
        )

    def main(self):
        try:
            dir_id = self._pick_uncached_directory()
        except NoDirectory:
            self.print_result("CRITICAL", "No directory exists in the archive.")
            return 2

        start_time = time.time()
        total_time = 0
        response = requests.post(self._url_for_dir(dir_id))
        assert response.status_code == 200, (response, response.text)
        result = response.json()
        while result["status"] in ("new", "pending"):
            time.sleep(self._poll_interval)
            response = requests.get(self._url_for_dir(dir_id))
            assert response.status_code == 200, (response, response.text)
            result = response.json()

            total_time = time.time() - start_time

            if total_time > self.critical_threshold:
                self.print_result(
                    "CRITICAL",
                    f"cooking directory {dir_id.hex()} took more than "
                    f"{total_time:.2f}s and has status: "
                    f'{result["progress_message"]}',
                    total_time=total_time,
                )

                self._collect_prometheus_metrics(2, total_time, ["cooking", "timeout"])

                return 2

        if result["status"] == "failed":
            self.print_result(
                "CRITICAL",
                f"cooking directory {dir_id.hex()} took {total_time:.2f}s "
                f'and failed with: {result["progress_message"]}',
                total_time=total_time,
            )

            self._collect_prometheus_metrics(2, total_time, ["cooking", "failed"])

            return 2
        elif result["status"] != "done":
            self.print_result(
                "CRITICAL",
                f"cooking directory {dir_id.hex()} took {total_time:.2f}s "
                f'and resulted in unknown status: {result["status"]}',
                total_time=total_time,
            )

            self._collect_prometheus_metrics(2, total_time, ["cooking", "unknown"])
            return 2

        (status_code, status) = self.get_status(total_time)

        if "fetch_url" not in result:
            self.print_result(
                "CRITICAL",
                f"cooking directory {dir_id.hex()} took {total_time:.2f}s "
                f"and succeeded, but API response did not contain a fetch_url.",
                total_time=total_time,
            )
            self._collect_prometheus_metrics(2, total_time, ["fetch", "no_url"])
            return 2

        with requests.get(result["fetch_url"], stream=True) as fetch_response:
            try:
                fetch_response.raise_for_status()
            except requests.HTTPError:
                self.print_result(
                    "CRITICAL",
                    f"cooking directory {dir_id.hex()} took {total_time:.2f}s "
                    f"and succeeded, but fetch failed with status code "
                    f"{fetch_response.status_code}.",
                    total_time=total_time,
                )
                self._collect_prometheus_metrics(2, total_time, ["fetch", "error"])
                return 2

            content_type = fetch_response.headers.get("Content-Type")
            if content_type not in ("application/gzip", "application/octet-stream"):
                self.print_result(
                    "CRITICAL",
                    f"Unexpected Content-Type when downloading bundle: {content_type}",
                    total_time=total_time,
                )
                self._collect_prometheus_metrics(
                    2, total_time, ["download", "unexpected_content_type"]
                )
                return 2

            try:
                with tarfile.open(fileobj=fetch_response.raw, mode="r|gz") as tf:
                    # Note that we are streaming the tarfile from the network,
                    # so we are allowed at most one pass on the tf object;
                    # and the sooner we close it the better.
                    # Fortunately, checking only the first member is good enough:
                    tarinfo = tf.next()
                    swhid = f"swh:1:dir:{dir_id.hex()}"
                    if not tarinfo or (
                        tarinfo.name != swhid
                        and not tarinfo.name.startswith(f"{swhid}/")
                    ):
                        self.print_result(
                            "CRITICAL",
                            f"Unexpected member in tarball: {tarinfo.name}"
                            if tarinfo
                            else "Fetched tarball is empty",
                            total_time=total_time,
                        )
                        self._collect_prometheus_metrics(
                            2, total_time, ["check", "archive_content"]
                        )
                        return 2
            except tarfile.ReadError as e:
                self.print_result(
                    "CRITICAL",
                    f"ReadError while reading tarball: {e}",
                    total_time=total_time,
                )
                self._collect_prometheus_metrics(
                    2, total_time, ["check", "archive_content"]
                )
                return 2
            except tarfile.StreamError as e:
                if (
                    sys.version_info < (3, 11)
                    and e.args[0] == "seeking backwards is not allowed"
                ):
                    # Probably https://github.com/python/cpython/issues/91078
                    self.print_result(
                        "CRITICAL",
                        f"StreamError while reading tarball (empty file?): {e}",
                        total_time=total_time,
                    )
                    self._collect_prometheus_metrics(
                        2, total_time, ["check", "archive_content"]
                    )
                    return 2

                self.print_result(
                    "CRITICAL",
                    f"StreamError while reading tarball: {e}",
                    total_time=total_time,
                )
                self._collect_prometheus_metrics(
                    2, total_time, ["check", "archive_content"]
                )
                return 2

        self.print_result(
            status,
            f"cooking directory {dir_id.hex()} took {total_time:.2f}s "
            f"and succeeded.",
            total_time=total_time,
        )

        self._collect_prometheus_metrics(status_code, total_time, ["end", ""])
        return status_code
